### PR TITLE
clean tmp file when index merge fail

### PR DESCRIPTION
**Pull request by binlijin**
Current when index merge fails, leave tmp file not cleaned.

### CODE REVIEW

**Comment by fjy** [comment]
@binlijin can u fix the spacing issues? also, does the problem exist in indexMerger v9?

File: processing/src/main/java/io/druid/segment/IndexMergerV9.java [hunk 1]
```diff
@@ -149,97 +149,113 @@ public Metadata apply(IndexableAdapter input)
...
-    ioPeon.cleanup();
-    FileUtils.deleteDirectory(v9TmpDir);
-    progress.stop();
+      v9Smoosher.close();
+      progress.stop();
 
-    return outDir;
+      return outDir;
+    } finally {
+      IOException e = null;
```
Comment (drcrallen) [inline processing/src/main/java/io/druid/segment/IndexMergerV9.java]: Can this just be done via a `Closer`?

The `Closer` code patterns have taken into account making sure everything closes and errors get thrown correctly.

Comment (binlijin) [inline processing/src/main/java/io/druid/segment/IndexMergerV9.java]: ok， will do.

**Comment by binlijin** [comment]
@fjy, ok, i think indexMergerV9 still have the  problem.

**Comment by binlijin** [comment]
@fjy  how to fix the spacing issues? I use intellij idea format the code.

**Comment by fjy** [comment]
@binlijin are you using hte style code in the repo? https://github.com/druid-io/druid/blob/master/intellij_formatting.jar

**Comment by binlijin** [comment]
@fjy, i am using the hte style code in the repo.

**Comment by fjy** [comment]
:+1:

**Comment by fjy** [comment]
@binlijin can you squash commits?

**Comment by jon-wei** [comment]
@binlijin 

Can we add a unit test that triggers the closing behavior?

**Comment by binlijin** [comment]
@jon-wei, how to test it? current i have no idea.

**Comment by jon-wei** [comment]
@binlijin 

I was thinking something like:

- Trigger an exception in IndexMerger, e.g. you can trigger an IllegalArgumentException in the IndexMergerTest test cases by falsely setting "hasSpatialIndexes" on a dimension in IncrementalIndex before persist() is called, e.g.:
```
    ColumnCapabilitiesImpl capabilities = (ColumnCapabilitiesImpl) toPersist.getCapabilities("dim1");
    capabilities.setHasSpatialIndexes(true);
```

- Verify that cleanup of tmpdir was successful
- Verify that cleanup of iopeon was successful (I'm fine leaving this out, since it looks like it would require an interface change to iopeon that lets you see the files that have been created)

**Comment by binlijin** [comment]
@jon-wei thanks,  add test for Closer.

**Comment by jon-wei** [comment]
:+1: after travis

**Comment by fjy** [comment]
:+1:

**Pull request merged**